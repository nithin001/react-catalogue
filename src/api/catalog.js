import instance from './base';
import Ajv from 'ajv';

export const get = () => new Promise((resolve, reject) => {
  instance.get('/catalog')
    .then((response) => {
      if (response.data === undefined) {
        reject('backend_error');
        return;
      }
      const data = response.data;
      const ajv = new Ajv();
      const validate = ajv.compile(schema);
      const valid = validate(data);
      if (valid) {
        resolve(data);
      } else {
        reject('backend_error');
      }
    })
    .catch((error) => {
      reject('backend_error');
    });
});

const schema = {
  $id: 'http://challenge.monoqi.net/catalog',
  type: 'object',
  definitions: {},
  $schema: 'http://json-schema.org/draft-07/schema#',
  additionalProperties: false,
  properties: {
    articles: {
      $id: '/properties/articles',
      type: 'array',
      items: {
        $id: '/properties/articles/items',
        type: 'object',
        additionalProperties: false,
        properties: {
          sku: {
            $id: '/properties/articles/items/properties/sku',
            type: 'string',
            title: 'The Sku Schema ',
            default: '',
          },
          name: {
            $id: '/properties/articles/items/properties/name',
            type: 'string',
            title: 'The Name Schema ',
            default: '',
          },
          image: {
            $id: '/properties/articles/items/properties/image',
            type: 'string',
            title: 'The Image Schema ',
            default: '',
          },
          price: {
            $id: '/properties/articles/items/properties/price',
            type: 'object',
            additionalProperties: false,
            properties: {
              amount: {
                $id: '/properties/articles/items/properties/price/properties/amount',
                type: 'number',
                title: 'The Amount Schema ',
                default: 0,
              },
              currency: {
                $id: '/properties/articles/items/properties/price/properties/currency',
                type: 'string',
                title: 'The Currency Schema ',
                default: '',
              },
            },
            required: [
              'amount',
              'currency',
            ],
          },
        },
        required: [
          'sku',
          'name',
          'image',
          'price',
        ],
      },
    },
  },
  required: [
    'articles',
  ],
};
