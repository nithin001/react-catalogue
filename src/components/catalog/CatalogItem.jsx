import React from 'react';
import { Link } from 'react-router-dom';
export const CatalogItem = ({ item, addToCart }) => (
  <div className={'catalog-item'}>
    <Link to={`/articles/${item.get('sku')}`}>
      <img src={item.get('image')} className={'catalog-item__image'}/>
    </Link>
    <span className={'catalog-item__name'}>
      <span className={'catalog-item__name catalog-item__name--legend'}>Name:&nbsp;</span>
      <span className={'catalog-item__name catalog-item__name--value'}>{item.get('name')}</span>
    </span>
    <br/>
    <span className={'catalog-item__price catalog-item__price--legend'}>Price: &nbsp;
      <span className={'catalog-item__price catalog-item__price--amount'}>{item.getIn(['price', 'amount'])}</span>
      &nbsp;
      <span className={'catalog-item__price catalog-item__price--currency'}>{item.getIn(['price', 'currency'])}</span>
    </span>
    <br/>
    <Link to={`/articles/${item.get('sku')}`}>
      <button className={'button button--details'}>View
      </button>
    </Link>
    <button className={'button button--add-to-cart'}
            onClick={() => {addToCart(item.get('sku'));}}>Add
    </button>
  </div>);

