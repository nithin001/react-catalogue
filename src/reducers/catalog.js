import Immutable from 'immutable';
const catalog = (state = Immutable.Map(), action) => {
  switch (action.type) {
    case 'POPULATE_CATALOG': {
      const articles = action.payload.articles;
      return Immutable.fromJS(articles);
    }
  }
  return state;
};
export default catalog;