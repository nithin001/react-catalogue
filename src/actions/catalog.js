import { get } from '../api/catalog';

export const load = () => (dispatch) => {
  dispatch({ type: 'LOAD_CATALOG_START' });
  return get()
    .then((catalog) => {
      dispatch({ type: 'POPULATE_CATALOG', payload: catalog });
      dispatch({ type: 'LOAD_CATALOG_COMPLETE' });
    })
    .catch(() => {
      dispatch({ type: 'LOAD_ERROR' });
    });
};
