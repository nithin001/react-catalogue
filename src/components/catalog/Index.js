import React from 'react';
import { connect } from 'react-redux';
import { CatalogItem } from './CatalogItem';
import { addToCart } from '../../actions/cart';
export const Catalog = ({ catalog, addToCart }) => {
  const articles = [];
  catalog.keySeq().forEach((item) => {
    articles.push(catalog.get(item));
  });
  const articlesList = articles.map(item => {
    return (<CatalogItem key={item.get('sku')} item={item} addToCart={addToCart}/>);
  });
  return (<div className={'catalog'}>
    {articlesList}
  </div>);
};

export const mapStateToProps = (state, ownProps) => {
  return { catalog: state.catalog };
};

export const mapDispatchToProp = (dispatch) => ({
    addToCart: (sku) => dispatch(addToCart(sku)),
  }
);

export default connect(mapStateToProps, mapDispatchToProp)(Catalog);


