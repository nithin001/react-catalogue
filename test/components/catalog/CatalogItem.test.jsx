import React from 'react';
import { Link } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { MemoryRouter } from 'react-router';
import { CatalogItem } from '../../../src/components/catalog/CatalogItem';
import catalog from '../../../src/reducers/catalog';

configure({ adapter: new Adapter() });
var mockCatalog = {
  'articles': [{
    'sku': '199203',
    'name': 'IPad',
    'image': 'https://picsum.photos/480/480/?image=6',
    'price': { 'amount': 229, 'currency': 'EUR' }
  }]
};

const catalogState = catalog(undefined, { type: 'POPULATE_CATALOG', payload: mockCatalog });
describe('CatalogItem component', () => {
  it('should render the item with name, image, price add to cart button', () => {
    const addToCart = jest.fn();
    const props = { item: catalogState.get(0), addToCart };
    const catalogItem = renderer.create(<MemoryRouter><CatalogItem
      {...props}
    /></MemoryRouter>);
    expect(catalogItem).toMatchSnapshot();
  });

  it('should call add to cart when button is clicked', () => {
    const addToCart = jest.fn();
    const props = { item: catalogState.get(0), addToCart };
    const catalogItem = shallow(<CatalogItem
      {...props}
    />);
    catalogItem.find(Link).at(0).simulate('click');
    expect(addToCart.mock.calls.length).toBe(1);
    expect(addToCart.mock.calls[0][0]).toEqual('199203');
  });
});