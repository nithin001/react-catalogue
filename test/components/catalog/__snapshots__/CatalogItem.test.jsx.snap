// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogItem component should render the item with name, image, price add to cart button 1`] = `
<div
  className="catalog-item"
>
  <a
    href="/articles/199203"
    onClick={[Function]}
  >
    <img
      className="catalog-item__image"
      src="https://picsum.photos/480/480/?image=6"
    />
  </a>
  <span
    className="catalog-item__name"
  >
    <span
      className="catalog-item__name catalog-item__name--legend"
    >
      Name: 
    </span>
    <span
      className="catalog-item__name catalog-item__name--value"
    >
      IPad
    </span>
  </span>
  <br />
  <span
    className="catalog-item__price"
  >
    <span
      className="catalog-item__price catalog-item__price--legend"
    >
      Price:  
    </span>
    <span
      className="catalog-item__price catalog-item__price--amount"
    >
      229
    </span>
     
    <span
      className="catalog-item__price catalog-item__price--currency"
    >
      EUR
    </span>
  </span>
  <br />
  <a
    href="/articles/199203"
    onClick={[Function]}
  >
    <button
      className="button"
    >
      View
    </button>
  </a>
  <button
    className="button button--inverted"
    onClick={[Function]}
  >
    Add
  </button>
</div>
`;
